### PR TITLE
Add test showing infinite unwinding

### PR DESCRIPTION
**Pull request by reuk**
In the attached test, CBMC tries to unwind the "recursive" virtual call forever, even though there's no actual recursion due to dynamic dispatch.

See diffblue/test-gen#845 to see why this is problematic in practice.

### CODE REVIEW

File: regression/cbmc-java/virtual_function_unwinding/test.desc [hunk 1]
```diff
@@ -0,0 +1,5 @@
+KNOWNBUG
+Virtual.class
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
```
Comment (thk123) [inline regression/cbmc-java/virtual_function_unwinding/test.desc]: Could you add a comment linking to the issue:

```
--
--
diffblue/test-gen#845
```

**Comment by reuk** [comment]
> is it important the bug that Outer implements Interface as well?
Yeah, the issue is that both `Inner.b` and `Outer.b` are the "same" function, so CBMC sees the nested call as recursion rather than virtual dispatch, which then gets unrolled forever.

**Comment by martin-cs** [comment]
Just as a general FYI (and please forgive me if it is patronising or
already known / obvious) : there will always be things like this.  This
one may be solvable but be careful how far down this rabbit hole you
want to go; it does not have a bottom and we have an approach (unwinding
assertions) for handling it.

During symex, you ideally want to eliminate dead branches and resolve
all manner of indirection (function pointers, etc.) in control flow.
However, due to the halting problem, this may be undecideable.  So symex
tends to do an over-approximate analysis (branch definitely not taken /
branch possibly taken).  This is done with constant folding mainly.
@tautschnig has a patch that forwards whole expressions and so can be
smarter about this.  I guess if you really wanted to go for it, you
could use the AI engine but (FWIW) I think unwinding assertions are a
better approach.

The purpose of unwinding assertions is to detect when a genuinely
reachable path has been cut by a depth / unwinding limit.  These can
then be used to extend the unwinding where it is needed, rather than
blindly which can get into this kind of loop.

HTH

**Comment by peterschrammel** [comment]
@reuk, please squash.

**Comment by reuk** [comment]
Squashed.

**Pull request merged**